dlt/destinations/sql_jobs.py

```diff
@@ -333,6 +337,44 @@ def gen_merge_sql(
                 )
             )
 
+        # remove "non-latest" records from staging table (deduplicate) if a sort column is provided
+        if len(primary_keys) > 0:
+            if has_column_with_prop(root_table, "sort"):
+                sort_column = escape_identifier(get_columns_names_with_prop(root_table, "sort")[0])
+                sql.append(f"""
+                    DELETE FROM {staging_root_table_name}
+                    WHERE {sort_column} IN (
+                        SELECT {sort_column} FROM (
+                            SELECT {sort_column}, ROW_NUMBER() OVER (partition BY {", ".join(primary_keys)} ORDER BY {sort_column} DESC) AS _rn
+                            FROM {staging_root_table_name}
+                        ) AS a
+                        WHERE a._rn > 1
+                    );
+                """)
+
+        # remove deleted records from staging tables if a hard_delete column is provided
+        if has_column_with_prop(root_table, "hard_delete"):
```

Comment:
I think (I hope) there's a simpler way to handle `hard_deletes`. The code below does not need any modifications. It will delete all rows from destination dataset (using primary and merge keys) that are present in the staging dataset. it does not matter if hard delete flag is set or not. we must delete those rows anyway.

we only must change how we insert, from here:
```
# insert from staging to dataset, truncate staging table
        for table in table_chain:
```
the only think you need to do is to filter out rows that have deleted flag set so this is another clause in where

Reply:
overall it should be way less code, we do not interfere with any edge cases by deleting and deduplicating the staging dataset + it looks like less row reads

Reply:
I have changed the approach to extending the where clause in the insert stage, rather than deleting from the staging dataset. It didn't turn out to be less code but it makes more sense nonetheless.